core_lib/src/structure/layervector.cpp

```diff
@@ -155,31 +155,53 @@ void LayerVector::loadDomElement(const QDomElement& element, QString dataDirPath
+        if (!imageElement.isNull() && imageElement.tagName() == "image")
         {
-            if (imageElement.tagName() == "image")
+            int position;
+            QString rawPath = imageElement.attribute("src");
+            if (!rawPath.isNull())
             {
-                int position;
-                if (!imageElement.attribute("src").isNull())
+                bool shouldLoad = true;
+
+                // Make sure src path is relative
+                shouldLoad &= QFileInfo(rawPath).isRelative();
+                QFileInfo fi(dataDirPath, rawPath);
+                // Make sure file is in data directory after resolving relative components and symlinks
+                QString canonicalPath = fi.canonicalFilePath();
+                fi.setFile(!canonicalPath.isEmpty() ? canonicalPath : fi.absoluteFilePath());
+                QDir dataDir(dataDirPath);
+                QDir ancestor = fi.dir();
+                while (ancestor != dataDir)
                 {
-                    QString path = dataDirPath + "/" + imageElement.attribute("src"); // the file is supposed to be in the data directory
-                    QFileInfo fi(path);
-                    if (!fi.exists()) path = imageElement.attribute("src");
-                    position = imageElement.attribute("frame").toInt();
-                    loadImageAtFrame(path, position);
+                    QDir newAncestor = QFileInfo(ancestor.absolutePath()).dir();
+                    if (ancestor == newAncestor)
+                    {
+                        // Data dir was not found in ancestors of the src path
+                        shouldLoad = false;
+                        break;
+                    }
+                    ancestor = newAncestor;
                 }
-                else
+
+                if (shouldLoad)
                 {
+                    QString path = fi.absoluteFilePath();
                     position = imageElement.attribute("frame").toInt();
-                    addNewKeyFrameAt(position);
-                    getVectorImageAtFrame(position)->loadDomElement(imageElement);
+                    loadImageAtFrame(path, position);
+                    getVectorImageAtFrame(position)->setOpacity(imageElement.attribute("opacity", "1.0").toDouble());
                 }
-                if (imageElement.hasAttribute("opacity"))
-                    getVectorImageAtFrame(position)->setOpacity(imageElement.attribute("opacity").toDouble());
-                else
-                    getVectorImageAtFrame(position)->setOpacity(1.0);
-                progressStep();
             }
+            else
+            {
+                position = imageElement.attribute("frame").toInt();
+                addNewKeyFrameAt(position);
+                getVectorImageAtFrame(position)->loadDomElement(imageElement);
+                getVectorImageAtFrame(position)->setOpacity(imageElement.attribute("opacity", "1.0").toDouble());
+            }
```

Comment:
I know that we had similar code but this feels wrong to me, why are we instantiating a keyframe if there's no src for it? Or let me rephrase, why do we create a new keyframe here at all when this method should only be about loading data if we have the necessary data.

It's not critique as much as a question, none of our other Layer classes have this so it seems weird to me that the Vector class has it, or do you disagree?

Reply:
It is weird to me that they are handled differently between layers yes. I did not change this behavior because it's harmless and I wanted to minimize the changes I make so it can be merged quicker and with lower risk of introducing a bug in this critical file loading code.

I don't think it is wrong to instantiate an empty frame if there is no src. If there is a bitmap key with no data, it does not save an image for that frame, but it still sets the src. For a vector key a .vec file is written and src is set even if there is no data. An src attribute pointing to a file that does not exist and a file that contains no data both feel wrong to me. If there's no data that needs to be stored for a key, we shouldn't need to create a data file. If there's no data file, we should not need to specify the path to the non-existent file. But users still may want to save an emtpy key (to stop holding the previous key, or for planning/timing purposes).